Recursive list

energyc0 opened the pull request:
Made recursive list library and recursive list exercise is done.

energyc0 committed:
Added recursive list library and recursive list exercise.

energyc0 committed:
Deleted useless printf()

WoWaster commented on 06.10.2025/recursiveList.h:
```diff
@@ -0,0 +1,58 @@
+ * Index may be greater than size of the list, but not less than 0.
+ * Return true if successfull.
+ */
+bool eraseRecursiveList(RecursiveList* pList, int idx);
```
Чаще `delete`, но пойдёт

WoWaster commented on 06.10.2025/recursiveList.h:
```diff
@@ -0,0 +1,58 @@
+typedef struct RecursiveList {
+    struct ListNode* _head;
+    int _size;
+} RecursiveList;
```
Давайте сделаем совсем красиво, сделаем здесь АТД и сокроем эту струтуру от пользователя. А ещё, `_` в начале названий полей для Си странно

WoWaster commented on 06.10.2025/recursiveList.c:
```diff
@@ -0,0 +1,185 @@
+
+    ListNode* node = pList->_head;
+    for (int i = 0; i < idx; i++, node = node->next)
+        ;
```
```suggestion
    for (int i = 0; i < idx; i++)
        node = node->next;
```

Давайте делать цикл `for` проще. И дальше по коду, если есть тоже.

WoWaster commented on 06.10.2025/recursiveList.c:
```diff
@@ -0,0 +1,185 @@
+        list._head = allocListNode(values[0], NULL, NULL);
+        if (list._head == NULL)
+            return list;
+        list._head->prev = list._head->next = tail = list._head;
```
> Сколько `=` Вы хотите? Да!

Буквы бесплатные. Лучше написать 3 строчки, которые сможет прочитать *любой*, а не только лишь самые упоротые Сишники.

WoWaster commented on 06.10.2025/recursiveList.c:
```diff
@@ -0,0 +1,185 @@
+    if (size > 0) {
+        list._head = allocListNode(values[0], NULL, NULL);
+        if (list._head == NULL)
+            return list;
```
А почему в остальных случаях Вы это не проверяете?

WoWaster commented on 06.10.2025/recursiveList.c:
```diff
@@ -0,0 +1,185 @@
+    return node;
+}
+
+RecursiveList makeRecursiveList(const int* values, int size)
```
Попробуйте сделать здесь логику чуть проще. Если вдруг размер 0, то просто верните `list` буквально сразу же. Тогда не будет `if` для головы и читатель не будет думать, почему цикл дальше не под `if`.

WoWaster commented on 06.10.2025/recursiveList.c:
```diff
@@ -0,0 +1,185 @@
+    if (pList->_head == NULL)
+        return;
+
+    for (ListNode* p = pList->_head;;) {
```
`while`?

WoWaster commented on 06.10.2025/recursiveList.c:
```diff
@@ -0,0 +1,185 @@
+    }
+}
+
+int getSizeRecursiveList(const RecursiveList* pList)
```
Хорошая заделка для АДТ, осталось запретить глупому пользователю трогать `_size` снаружи.

WoWaster commented on 06.10.2025/recursiveList_exercise.c:
```diff
@@ -0,0 +1,63 @@
+
+    char buf[BUFSIZ];
+    while (fgets(buf, sizeof(buf) - 1, stdin) != NULL
+        && (res = sscanf(buf, "%d", num)) < 1) {
```
А чем не нравится тупой `scanf`?

WoWaster commented on 06.10.2025/recursiveList_exercise.c:
```diff
@@ -0,0 +1,63 @@
+    char buf[BUFSIZ];
+    while (fgets(buf, sizeof(buf) - 1, stdin) != NULL
+        && (res = sscanf(buf, "%d", num)) < 1) {
+        if (res == -1)
```
И правда ли `sscanf` может вернуть `-1`?

energyc0 replied:
Да, когда строка пустая, то есть пользователь мог ввести EOF. Хотя, тут fgets() уже проверяет это, поэтому я с Вами согласен!

WoWaster commented on 06.10.2025/recursiveList_exercise.c:
```diff
@@ -0,0 +1,63 @@
+ */
+static int countingGame(int n, int m)
+{
+    int values[n];
```
Надо инициализировать

energyc0 replied:
```bash
recursiveList_exercise.c: In function ‘countingGame’:
recursiveList_exercise.c:32:21: warning: ISO C forbids empty initializer braces before C23 [-Wpedantic]
   32 |     int values[n] = {};
      |                  
```

WoWaster commented on 06.10.2025/recursiveList_exercise.c:
```diff
@@ -0,0 +1,63 @@
+
+int main(void)
+{
+    int n;
```
И это тоже.

energyc0 replied:
Но я же инициализирую это значение далее, а потом возвращаю код возврата 1, если не удалось.

WoWaster replied:
Я понимаю, что Вы понимаете. Мысль за этими словами простая -- добиться мышечной памяти в плане инициализации значений.

WoWaster commented on 06.10.2025/sortedList_exercise.c:
А это откуда здесь?!

WoWaster reviewed (changes requested):
3 балла.

energyc0 committed:
Fixed bad code style.
Initialize variables, incapsulate recursive list in recursiveList.h, fixed bad code style.

energyc0 commented:
Cleaned the history, deleted all info about sorted list.
